torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -327,6 +327,22 @@ class FullyShardedDataParallel(nn.Module):
             avoid sharding specific parameters when using an
             ``auto_wrap_policy`` or if parameters' sharding is not managed by
             FSDP. (Default: ``None``)
+        param_init_fn: (Optional[Callable[[nn.Module], None]]):
+            A ``Callable[torch.nn.Module] -> None`` that
+            specifies how modules that are currently on the meta device should be initialized
+            onto a actual device. Note that if this argument is passed in, it is assumed that
+            the corresponding module is currently on the meta device.
+            The same ``Callable`` is applied to initialize all meta modules. Note that this
+            initialization function is applied before doing any FSDP sharding
+            logic.
+
+            Example::
+
+                >>> module = MyModule(device="meta")
+                >>> def my_init_fn(module):
+                >>>     # responsible for initializing a module, such as with reset_parameters
+                >>> fsdp_model = FSDP(module, param_init_fn=my_init_fn, auto_wrap_policy=default_auto_wrap_policy)
```

Comment:
Would I be correct that we don't need this if after we land deferred init to PyTorch repo, as that will capture the init behaviors automatically?

Reply:
We could do that if we decide to use deferred_init + materialize_module API as the one offering for meta device init in PT when it is available. 

On the other hand, we might still want to keep `param_init_fn` if user wants to test out some custom initialization logic?

Reply:
Refactored to capture some init behavior automatically after discussion with @zhaojuanmao